src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -52,9 +52,17 @@ public partial class JsonTypeInfo<T>
                 WriteStack state = default;
                 state.Initialize(this, rootValueBoxed);
 
-                bool success = EffectiveConverter.WriteCore(writer, rootValue, Options, ref state);
-                Debug.Assert(success);
-                writer.Flush();
+                try
+                {
+                    bool success = EffectiveConverter.WriteCore(writer, rootValue, Options, ref state);
+                    Debug.Assert(success);
+                    writer.Flush();
+                }
+                catch
+                {
+                    state.DisposePendingDisposablesOnException();
```

Comment:
I would just move this logic inside the `WriteCore` method, which already has a `try`/`catch`. It should check if `state.SupportAsync` is false first, because the async serializer is already calling the equivalent async disposal method.

Reply:
I would recommend to keep the current proposal because:
- it is better, when possible, to dispose a resource, here the `WriteStack`, in the same scope as it is created,
- it is consistent with the existing async serialization flow,
- it would add another (global) try/catch block in the `WriteCore` method to ensure all exceptions are correctly handled.
Do I miss something?